Consolidate the naming of named_parameter and state_dict for CheckpointWrapper

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80089

named_parameter() should return the same parameter names as state_dict() but the current CheckpointWrapper does not enforce this naming rule. This PR resolves this issue.

Differential Revision: [D37344200](https://our.internmc.facebook.com/intern/diff/D37344200/)

fegin committed:
Consolidate the naming of named_parameter and state_dict for CheckpointWrapper

named_parameter() should return the same parameter names as state_dict() but the current CheckpointWrapper does not enforce this naming rule. This PR resolves this issue.

Differential Revision: [D37344200](https://our.internmc.facebook.com/intern/diff/D37344200/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -61,6 +61,19 @@ def forward(self, *args, **kwargs):
+        Overrides :meth:`named_parameters()` to intercept parameter names and
+        remove all occurrences of _CHECKPOINT_PREFIX.
+        """
+        # Determine which logic to use based on the context at call time
```
nit: there is no context here?

rohan-varma reviewed (commented):
lgtm, thanks

fegin committed:
Update on "Consolidate the naming of named_parameter and state_dict for CheckpointWrapper"


named_parameter() should return the same parameter names as state_dict() but the current CheckpointWrapper does not enforce this naming rule. This PR resolves this issue.

Differential Revision: [D37344200](https://our.internmc.facebook.com/intern/diff/D37344200/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
lgtm

fegin committed:
Update on "Consolidate the naming of named_parameter and state_dict for CheckpointWrapper"


named_parameter() should return the same parameter names as state_dict() but the current CheckpointWrapper does not enforce this naming rule. This PR resolves this issue.

Differential Revision: [D37344200](https://our.internmc.facebook.com/intern/diff/D37344200/)

[ghstack-poisoned]

fegin committed:
Update on "Consolidate the naming of named_parameter and state_dict for CheckpointWrapper"


named_parameter() should return the same parameter names as state_dict() but the current CheckpointWrapper does not enforce this naming rule. This PR resolves this issue.

Differential Revision: [D37344200](https://our.internmc.facebook.com/intern/diff/D37344200/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
